packages/portalnetwork/src/wire/utp/Socket/UtpSocket.ts

```diff
@@ -87,213 +121,198 @@ export class UtpSocket extends EventEmitter {
 
-  async sendSynPacket(packet: Packet): Promise<ConnectionState> {
-    await this.sendPacket(packet, PacketType.ST_SYN)
-    this.state = ConnectionState.SynSent
-    return this.state
+  createPacket<T extends PacketType>(
+    opts: ICreatePacketOpts<T> = {} as ICreatePacketOpts<T>
+  ): Packet<T> {
+    if ('bitmask' in opts && 'payload' in opts) {
+      throw new Error('Bitmask and payload are mutually exclusive')
+    }
+    if ('payload' in opts && opts.pType !== PacketType.ST_DATA) {
+      throw new Error('Payload can only be set for data packets')
+    }
+    opts.pType === PacketType.ST_DATA && this.seqNr++
+    const extension = 'bitmask' in opts ? 1 : 0
+    const params: ICreate<T> = {
+      ...opts,
+      seqNr: this.seqNr,
+      ackNr: this.ackNr,
+      extension,
+    }
+    return this.packetManager.createPacket<T>(params)
   }
 
-  async sendSynAckPacket(packet: Packet): Promise<void> {
-    await this.sendPacket(packet, PacketType.ST_STATE)
+  async sendSynPacket(): Promise<void> {
+    const p = this.createPacket<PacketType.ST_SYN>({ pType: PacketType.ST_SYN })
+    await this.sendPacket<PacketType.ST_SYN>(p, PacketType.ST_SYN)
+    this.state = ConnectionState.SynSent
   }
-
-  async sendDataPacket(packet: Packet): Promise<Packet> {
-    this.state = ConnectionState.Connected
-    this.outBuffer.set(packet.header.seqNr, packet.header.timestampMicroseconds)
-    this.cur_window = this.outBuffer.size * DEFAULT_PACKET_SIZE
-    await this.sendPacket(packet, PacketType.ST_DATA)
-    this.seqNr++
-    return packet
+  async sendAckPacket(bitmask?: Uint8Array): Promise<void> {
+    const packet = bitmask
+      ? this.createPacket<PacketType.ST_STATE>({ pType: PacketType.ST_STATE, bitmask })
+      : this.createPacket<PacketType.ST_STATE>({ pType: PacketType.ST_STATE })
+    await this.sendPacket<PacketType.ST_STATE>(packet, PacketType.ST_STATE)
   }
-
-  async sendStatePacket(packet: Packet): Promise<void> {
-    await this.sendPacket(packet, PacketType.ST_STATE)
+  async sendSynAckPacket(): Promise<void> {
+    this.state = ConnectionState.SynRecv
+    await this.sendAckPacket()
   }
-
-  async sendResetPacket(packet: Packet) {
+  async sendResetPacket() {
     this.state = ConnectionState.Reset
-    await this.sendPacket(packet, PacketType.ST_RESET)
+    const packet = this.createPacket<PacketType.ST_RESET>({ pType: PacketType.ST_RESET })
+    await this.sendPacket<PacketType.ST_RESET>(packet, PacketType.ST_RESET)
   }
 
-  async sendFinPacket(packet: Packet): Promise<Buffer> {
+  async sendFinPacket(): Promise<void> {
+    const packet = this.createPacket<PacketType.ST_FIN>({ pType: PacketType.ST_FIN })
     this.finNr = packet.header.seqNr
-    return await this.sendPacket(packet, PacketType.ST_FIN)
+    await this.sendPacket<PacketType.ST_FIN>(packet, PacketType.ST_FIN)
+  }
+
+  async sendDataPacket(bytes: Uint8Array): Promise<Packet<PacketType.ST_DATA>> {
+    await this.packetManager.congestionControl.canSend()
```

Comment:
Don't we need to check the result of this call before sending a packet?  Out of curiosity, how does the usage of this function differ from the `canSend` event?